### ISSUE TITLE
daemonset status.numberReady does not correspond to actual daemonset pod state

### ISSUE THREAD

**Issue by richm** [kubernetes/kubernetes#51278]
/kind bug

**What happened**:
Using a daemonset pod.  The pod is not running because the node is not labeled as per the node selector.  Label the node.  Query the daemonset information.  status.numberReady is reported as 1, but it may take many seconds for the pod to be in the Running state.  Same with shutdown.  Unlabel the node.  Query the daemonset information.  status.numberReady is reported as 0, but the pod may be in the Terminating state for many seconds.

**What you expected to happen**:

The status.numberReady does not report 1 until the pod is actually in the Running state at start up.  The status.numberReady does not report 0 until the pod is terminated and removed at shutdown.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.1+5115d708d7", GitCommit:"fff65cf", GitTreeState:"clean", BuildDate:"2017-08-02T21:42:22Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.1+5115d708d7", GitCommit:"fff65cf", GitTreeState:"clean", BuildDate:"2017-08-02T21:42:22Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}

- Cloud provider or hardware configuration**:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by richm** [comment]
@kubernetes/sig-node-bugs

**Comment by stevekuznetsov** [comment]
@kubernetes/sig-apps-bugs

**Comment by enisoc** [comment]
cc @kow3ns @janetkuo 

On the terminating side, #51279 might help.

**Comment by janetkuo** [comment]
Terminating pods would not be ready anymore at some point before they're fully terminated. Marking `status.numberReady` as 0 is expected.

**Comment by stevekuznetsov** [comment]
@janetkuo I agree that considering a Terminating pod as not ready makes sense -- it's the obverse (Query the daemonset information. status.numberReady is reported as 1, but it may take many seconds for the pod to be in the Running state.) that is concerning

**Comment by janetkuo** [comment]
@stevekuznetsov When DaemonSet `.status.numberReady = 1`, is the pod ready (check pod's `.status.coditions` to see if there's one pod condition whose `type=Ready` and `status=True`)?

DaemonSet checks pod's ready condition and then updates DaemonSet's `.status.numberReady` accordingly. A pod's Ready condition and Running state are updated separately in the system. I suspect that caused the delay you observed.

**Comment by stevekuznetsov** [comment]
@richm could you provide some timed logs for the behavior you noticed so we can confirm we're seeing the behavior @janetkuo suggests is correct?

**Comment by richm** [comment]
I won't be able to do this any time soon.

**Comment by zouyee** [comment]
![2018-05-25 3 45 25](https://user-images.githubusercontent.com/5785605/40532628-ab43fc38-6032-11e8-81ab-4a3be3321542.png)
